Add support for Objective-C classes

jacob-carlborg opened the pull request:
This PR contains a bit more code then I would like but it's difficult to reduce it anymore.

The PR consists of four commits and can be individually reviewed but doesn't make sense to split up in multiple PRs.

Spec update: https://github.com/dlang/dlang.org/pull/2540.

jacob-carlborg committed:
Add support for Objective-C classes

jacob-carlborg committed:
Add ABI documentation for Objective-C classes

jacob-carlborg committed:
Add internal support for anonymous variable declarations

jacob-carlborg committed:
Deprecate Objective-C interfaces

Representing an Objective-C class as a D interface has been
deprecated. Classes have now been properly implemented and the `class`
keyword should be used instead.

In the future, `extern(Objective-C)` interfaces will be used to
represent Objective-C protocols.

Geod24 commented on changelog/objc_class.dd:
```diff
@@ -0,0 +1,49 @@
+$(UL
+    $(LI `static` and `final` methods are virtual. Although `final` methods are
+         virtual it's not possible to override a `final` method in a subclass)
+    $(LI `static` methods are overridable in subclasses)
```
That sounds quite fishy. An `extern` declaration is used for compatibility, but we always stayed away from importing other language's semantic into D, as it's a hornets nest. It's fine to have `final` method as virtual if ObjC has no way to represent it, but allowing `override static` is downright confusing.

I assume this comes from the necessity to have it in the vtable, even if its static ?

jacob-carlborg replied:
Class/static methods in Objective-C or instance methods on the metaclass. In Objective-C classes are object themselves.

jacob-carlborg replied:
> I assume this comes from the necessity to have it in the vtable, even if its static ?

There is no vtable for Objective-C classes. All method calls are lowered to a function in the Objective-C runtime.

adamdruppe replied:
Without this, it is basically impossible to actually use existing Objective C classes. We sometimes need to be practical even when it sounds weird...

Geod24 replied:
@jacob-carlborg : Is there a way to expose this without changing the language ?

jacob-carlborg replied:
Expose what? That static methods are overridable?

Geod24 commented:
Using the HEAD of this PR (`DMD64 D Compiler v2.084.0-rc.1-161-g4b99aba26) and slightly modifying the example code:

```D
void main()
{
    auto foo = Foo.alloc.init;
    scope (exit) foo.release();

    assert(foo.bar(3) == 3);

    pragma(msg, &__traits(getOverloads, Foo, "alloc"));
}

extern (Objective-C):
class NSObject
{
     static NSObject alloc() @selector("alloc");
     NSObject init() @selector("init");
     void release() @selector("release");
 }

class Foo : NSObject
{
    override static Foo alloc() @selector("alloc");
    override Foo init() @selector("init");

    int bar(int a) @selector("bar:")
    {
        return a;
    }
}
```

```
~/projects/dmd (git)-[remotes/jacob/objc-class] % ./generated/osx/release/64/dmd -run test.d                 :(
---
ERROR: This is a compiler bug.
Please report it via https://issues.dlang.org/enter_bug.cgi
with, preferably, a reduced, reproducible example and the information below.
DustMite (https://github.com/CyberShadow/DustMite/wiki) can help with the reduction.
---
DMD v2.084.0-rc.1-161-g4b99aba26
predefs   DigitalMars Posix OSX CppRuntime_Clang darwin LittleEndian D_Version2 all D_SIMD D_InlineAsm_X86_64 X86_64 D_LP64 D_PIC assert D_ModuleInfo D_Exceptions D_TypeInfo D_HardFloat D_ObjectiveC
binary    ./generated/osx/release/64/dmd
version   v2.084.0-rc.1-161-g4b99aba26
config    ./generated/osx/release/64/dmd.conf
DFLAGS    -I./generated/osx/release/64/../../../../../druntime/import -I./generated/osx/release/64/../../../../../phobos -L-L./generated/osx/release/64/../../../../../phobos/generated/osx/release/64 -fPIC
---
core.exception.AssertError@dmd/hdrgen.d(2300): Assertion failure
----------------
??:? _d_assertp [0xb92f689]
dmd/access.d:623 _ZN18PrettyPrintVisitor11expToBufferEP10Expression4PREC [0xb795831]
dmd/access.d:623 _ZN18PrettyPrintVisitor5visitEP9DotVarExp [0xb79778d]
dmd/access.d:623 _ZN9DotVarExp6acceptEP7Visitor [0xb75d395]
[...]
```

WalterBright commented:
> doc/objectivec-abi.md

The official documentation is in Ddoc .dd files. Trying to mix and match .md and .dd files will cause a lot of bugs and effort for others trying to build the dlang website.

WalterBright commented on src/dmd/declaration.d:
```diff
@@ -1097,11 +1097,18 @@ extern (C++) class VarDeclaration : Declaration
 
     VarDeclarations* maybes;        // STC.maybescope variables that are assigned to this STC.maybescope variable
 
+    private bool isAnonymous_;
```
What does this do? Why does it have a trailing _ ?

jacob-carlborg replied:
It has a trailing `_` to avoid a conflict with the method `isAnonymous`. It indicates that variable is anonymous that therefore will not be added to the symbol table.

ZombineDev replied:
The [convention](https://dlang.org/dstyle#naming_keywords) is to use `name_` (instead of just `name`) when `name` is a keyword. If the variable is aggregate-level (module, class, etc.) private it [can be prefixed with an underscore](https://dlang.org/dstyle#naming_general).
Since the `_name` style is used in other languages (in addition to being allowed by the style guide) I suggest:
```suggestion
    private bool _isAnonymous;
```

jacob-carlborg replied:
Fine, but let the record show I do this change under protest 😃.

jacob-carlborg replied:
Fixed.

wilzbach commented:
> > doc/objectivec-abi.md
> 
> The official documentation is in Ddoc .dd files. Trying to mix and match .md and .dd files will cause a lot of bugs and effort for others trying to build the dlang website.

These files aren't part of the official user documentation, but for developers and in markdown as that's what GitHub understands when you browse those files on the web (which only DMD devs or very interested people would do).

WalterBright commented on src/dmd/func.d:
```diff
@@ -198,6 +198,22 @@ extern (C++) struct Ensure
+         * method is called on. For nested functions it would be the enclosing
+         * function's stack frame.
+         */
+        VarDeclaration this_;
```
Why this_ instead of vthis (as used elsewhere)?

WalterBright replied:
And how is it different from .vthis?

jacob-carlborg replied:
It is the same as `vthis`, I can renamed it but I think `this_` is a better name and follows D naming conventions.

WalterBright replied:
The convention for a variable representing the `this` value is `vthis`. This appears in the front end 146 times.

jacob-carlborg replied:
Doesn't mean it's a good name, but I'll rename it.

jacob-carlborg replied:
Renamed.

WalterBright commented on src/dmd/func.d:
```diff
@@ -220,6 +236,7 @@ extern (C++) class FuncDeclaration : Declaration
     VarDeclaration vthis;               /// 'this' parameter (member and nested)
     VarDeclaration v_arguments;         /// '_arguments' parameter
     ObjcSelector* selector;             /// Objective-C method selector (member function only)
+    VarDeclaration selectorParameter;   /// Objective-C implicit selector parameter
```
How is this different from HiddenParameters.selector?

jacob-carlborg replied:
It's the same. `HiddenParameters` is just a way to return multiple values from `declareThis`. Since D doesn't have built-in way to return tuples and we're not allowed to use Phobos I added `HiddenParameters`. I named it `HiddenParameters.selector` and not `HiddenParameters.selectorParameter` because `HiddenParameters` already implies it's a parameter. But I can rename it if you prefer.

jacob-carlborg replied:
Renamed `HiddenParameters.selector` to `HiddenParameters.selectorParameter`.

WalterBright commented on src/dmd/identifier.d:
```diff
@@ -30,6 +30,8 @@ import dmd.utils;
 extern (C++) final class Identifier : RootObject
 {
 private:
+    __gshared Identifier anonymous_;
```
We have `anonymous`, `anonymous_`, and `__anonymous`. Need more helpful names.

jacob-carlborg replied:
Not sure I understand. `anonymous` is the name of the getter and I cannot have the same name of a method as an instance variable so I added the underscore, standard D naming conventions.

jacob-carlborg replied:
This has been implicitly fixed by fixing other comments.

WalterBright commented on src/dmd/identifier.d:
```diff
@@ -66,6 +68,17 @@ public:
+        anonymous_ = new Identifier("__anonymous", TOK.identifier);
+    }
+
+    /// Placeholder for an anonymous identifier.
```
It isn't a placeholder. It's a singleton. The code could be simplified and made clearer by moving the initialization of `anonymous_` and the `__gshared` variable into this function and making it initialize lazilly.

jacob-carlborg replied:
That would cause potential issues with threads. But if we don't care about that I can make the change.

thewilsonator replied:
What Walter is saying is this should either be a `generateId` and thus be local, or it should be part of [`Id`](https://github.com/dlang/dmd/blob/master/src/dmd/id.d#L26) like all the other compiler recognised/generated ones.

WalterBright replied:
The front end is single threaded so no issue with threads. It could be made multithreaded by using `static` instead of `__gshared`, but there's no reason to do that at this time.

jacob-carlborg replied:
> What Walter is saying is this should either be a generateId and thus be local, or it should be part of Id like all the other compiler recognised/generated ones.

It is generated locally. It's intended usage is like this:

```d
void foo(Identifier ident)
{
    if (ident is Identifier.anonymous) // hence why it's a placeholder
        ident = new Identifier(...);
}
```

thewilsonator replied:
So its a sentinel. Can you get away with using `null`?

jacob-carlborg replied:
> So its a sentinel.

That might be a better name.

> Can you get away with using null?

Yes, but:

* It's not descriptive/clear what's going on. I prefer encoding the intent in the actual code than in comments
* There's nothing stopping `ident` from being `null` for other reasons

thewilsonator replied:
>There's nothing stopping ident from being null for other reasons

I was hoping that was not going to be the case, oh well.

jacob-carlborg replied:
@thewilsonator another approach than using a sentinel would be to add a new field to `Identifier` to indicate it's anonymous. This would have the advantage of allowing to add a more descriptive name to an identifier that can help with debugging.

WalterBright commented on src/dmd/mars.d:
```diff
@@ -336,6 +336,7 @@ private int tryMain(size_t argc, const(char)** argv)
     // Initialization
     Type._init();
     Id.initialize();
+    Identifier.initialize();
```
This wouldn't be needed if the singleton was used as mentioned above.

WalterBright commented on src/dmd/objc.d:
```diff
@@ -319,12 +465,55 @@ extern(C++) private final class Supported : Objc
+            });
+
+            return result;
+        }
```
You can eliminate the `result` variable by just checking `foreachDsymbol` for a non-zero result.

jacob-carlborg replied:
Ok, but I think the code will be more difficult to read then.

jacob-carlborg replied:
Fixed.

WalterBright commented on src/dmd/objc.d:
```diff
@@ -319,12 +465,55 @@ extern(C++) private final class Supported : Objc
+
+            members.foreachDsymbol((member) {
+                if (auto attrib = member.isAttribDeclaration)
+                    result = isExtern(attrib.decl);
```
This looks like a bug, because it does not stop iteration, and may get set to something different on the next iteration.

jacob-carlborg replied:
Hmm, yes.

jacob-carlborg replied:
Fixed.

WalterBright commented on src/dmd/objc.d:
```diff
@@ -162,10 +191,67 @@ extern(C++) abstract class Objc
+     * Returns: the parent
+     */
+    abstract ClassDeclaration getParent(FuncDeclaration functionDeclaration,
+        ClassDeclaration classDeclaration) const;
```
Long names imply a global name, short names imply locals. These are locals.

> cdecl is used throughout the DMD source code. 

I can't find a single instance of it. Commonly used dmd local names are:

1. ad = Aggregate Declaration
2. sd = StructDeclaration
3. cd = ClassDeclaration

This convention makes code much easier to read.

jacob-carlborg replied:
> Long names imply a global name, short names imply locals. These are locals.

Long names imply descriptive names. I used the descriptive names in the abstract base class because those are the names that will be visible when the documentation is generated. For an outside user it's going to be much more clear. For anyone already familiar with the specific DMD naming conventions, I don't see how it can be any confusing, because `classDeclaratiaon` cannot mean anything else than "class declaration".

But fine, I'll rename it.

jacob-carlborg replied:
Renamed.

jacob-carlborg commented:
> The ABI is part of the spec and should be on the website with the spec, just like the rest of the D ABI documentation. https://dlang.org/spec/abi.html

We don't have any documentation for the ABI for C or C++ on dlang.org, why would we for Objective-C?

jacob-carlborg commented on src/dmd/objc.d:
```diff
@@ -162,10 +191,67 @@ extern(C++) abstract class Objc
+     *
+     * Returns: `true` if the given function declaration is virtual
+     */
+    abstract bool isVirtual(const FuncDeclaration functionDeclaration) const;
```
Renamed.

jacob-carlborg commented:
I've fixed most things.

WalterBright commented:
> We don't have any documentation for the ABI for C or C++ on dlang.org, why would we for Objective-C?

1. You felt there was a need to specify how it works.
2. You already wrote it.
3. Why not use it?
4. Why hide it on github?

jacob-carlborg commented:
> This good to go?

No, I need to fix the following two comments: https://github.com/dlang/dmd/pull/9179#issuecomment-451031022 and https://github.com/dlang/dmd/pull/9179#discussion_r244898079. I'll put it as a WIP until it's fixed.

thewilsonator commented:
Thanks.

jacob-carlborg commented:
@Geod24 that error in https://github.com/dlang/dmd/pull/9179#issuecomment-451031022 is not something new in this PR. It exists in current DMD. Well, it segfaults rather than asserts.

jacob-carlborg commented on src/dmd/semantic2.d:
```diff
@@ -440,7 +440,9 @@ private extern(C++) final class Semantic2Visitor : Visitor
         objc.validateSelector(fd);
         if (ClassDeclaration cd = fd.parent.isClassDeclaration())
         {
+            cd = objc.getParent(fd, cd);
```
Fixed. I move the all these Objective-C specific semantic functions to the first semantic phase, where the original were located.

jacob-carlborg commented:
I've fixed the https://github.com/dlang/dmd/pull/9179#discussion_r244898079 comment now. I moved  all the Objective-C specific semantic functions for function declaration to the first semantic phase, where the original were located.

I'll leave the issue that @Geod24 mentioned because it was already present before this PR. It doesn't look like it's possible to take the address of a method in Objective-C. It's possible to replace methods at runtime in Objective-C, but the Objective-C runtime provides functions to get access to the actual implementation. Because of this I think we should make it illegal to take the address of a function with Objective-C linkage in D.

jacob-carlborg commented:
@thewilsonator This is good to go, unless I should change the implementation of anonymous variable declarations, see: https://github.com/dlang/dmd/pull/9179#discussion_r245520430.

thewilsonator commented:
OSX64 auto-tester failed with
```
 ... fail_compilation/objc_class3.d  -fPIC ()
Test fail_compilation/objc_class3.d failed.  The logged output:
/Users/braddr/sandbox/at-client/pull-3472605-Darwin_64_64/dmd/generated/osx/release/64/dmd -conf= -m64 -Ifail_compilation -verrors=0  -fPIC  -odgenerated/fail_compilation -ofgenerated/fail_compilation/objc_class3_0.o  -c fail_compilation/objc_class3.d 


==============================
Test fail_compilation/objc_class3.d failed: expected rc == 1, but exited with rc == 0

make[2]: *** [generated/fail_compilation/objc_class3.d.out] Error 1
```

Not particularly helpful message.

jacob-carlborg commented:
> Not particularly helpful message.

It's a test that should fail to compile to succeeds. Hence the unhelpful message.

jacob-carlborg commented:
Looks like the tests are passing now.

thewilsonator commented:
Auto-merge toggled on

wilzbach commented:
> Auto-merge toggled on

Why don't you use "auto-merge"?
It _works_ very reliably and actually doesn't merge it automatically when there's a CI failure as these should always be inspected manually (or be avoided in the first place).

jacob-carlborg commented:
> Why don't you use "auto-merge"?

Seems he triggered it in the auto-tester UI and not here on GitHub.

thewilsonator commented:
Bingo. 

I need to upgrade my system because GH has dropped support for it and broke almost everything in the process.

wilzbach commented:
> I need to upgrade my system because GH has dropped support for it and broke almost everything in the process.

So you're saying that your Chrome/Firefox doesn't allow you to add this anymore:

![image](https://user-images.githubusercontent.com/4370550/50830228-81a21680-1347-11e9-933a-28491d5008c4.png)

FWIW it is still just a one-liner with their API and curl:

```
curl -u "wilzbach:$GITHUB_OAUTH_TOKEN" --request POST --data '{"labels": ["auto-merge"]}' https://api.github.com/repos/dlang/dmd/issues/9179/labels
```
